### ISSUE TITLE
nxagent drops X client connections unexpectedly

### ISSUE THREAD

**Issue by mbecroft** [ArcticaProject/nx-libs#589]
Version: release 3.5.99.10 built from release tarball

Using nx in 'virtual desktop' mode.

From time to time, nxagent will unexpectedly close the connection to an X client. The result is that client's windows disappear, and the client emits an error such as: `Fatal IO error 11 (Resource temporarily unavailable) on X server :124.`

nxagent itself, the connection from the NX client to nxagent (if connected at the time of the fault), and all other clients continue working normally.

The client program that failed can be restarted by the user immediately, and continues working normally...until the next time the problem reoccurs.

The problem seems to happen mostly with 'heavy' clients that own a large number of resources as reported by `xrestop`.

Specifically, I experience the fault frequently with Firefox 38 ESR and Palemoon web browsers, and occasionally with the fvwm2 window manager. It has never happened with: emacs, xterm, xclock.

In the case of the web browser (Firefox or Palemoon) the browser will die seemingly at random...sometimes while I am not interacting with the system at all, other times when I am in the middle of browsing to a web site. There is no obvious pattern except that sometimes the browser will die (and I will restart it) every 10 minutes or so; other times it can run for several days...but always will eventually die with this fault.

Often I have a large number of tabs in the browser with correspondingly high resource utilisation. But the problem can also happen when I have just a few windows open.

No relevant messages appear on the browser stderr (except the error noted above) nor in the nxagent stderr found in (for instance) `/tmp/.x2go-mb/C-mb-124-1508677854_stD.xsession-x2go_dp24/session.log`.

In all other respects there are no problems. I am able to suspend and resume NX sessions as normal. The NX session never dies or closes unexpectedly.

In this particular case, the X clients experiencing intermittent disconnection are running on the same host as nxagent and connecting to nxagent via UNIX sockets. I have not tested whether the problem also occurs with clients connecting via TCP sockets.

I am challenged by how to better characterize or analyze the issue.

This issue corresponds with a discussion in `#x2go` around 20171205T0000 UTC in which I (`mbecroft`) explained the issue and `uli42` made some comments and suggested raising this ticket.

**Comment by mbecroft** [comment]
*Further observation* - the above described issue gradually gets worse and worse the longer nxagent has been running over a period of weeks (even if individual X clients have been closed and restarted many times).

*Further issue - possibly (probably) related.*

What I've also observed is that in addition to just dropping clients, things will start to go slower and slower, until ridiculously slow. But neither nxagent itself nor anything else is CPU-bound, nor is there any disk I/O.

I can max out nxagent at 100% CPU easily by doing graphics intensive tests when it is running well.

But when in 'go slow' state, it will be sitting at maybe 10% CPU and can't do any more than that.

Invariably, there will be one X client (today it was libreoffice) which is causing the problem, and the moment I quit that client, everything comes back to normal.

**Comment by mbecroft** [comment]
Eventually today nxagent crashed completely. May or may not be in any way related. Backtrace can be found here: https://pastebin.com/hpk9LRZ5

I am rebuilding from current arctica git 3.6.x HEAD (804ff44523ab726fdb59b80193652465df0482ee) and will test that for a while. It looks like uli42's memory leak fixed have just made it in time for this, so hoepfully that may help.

It also looks like there might be something not quite right with my libfreetype so I'm going to remerge that as well. Maybe, here's hoping, whatever the bug is has been fixed. Will update this ticket in a few days with any observations.

**Comment by uli42** [comment]
maybe it is related to #586

**Comment by uli42** [comment]
There reference to Font.c refers to libNX_X11's Font.c, not the one from nxagent. And that code looks like this:
```
    166 #ifdef USE_XF86BIGFONT
    167         _XF86BigfontFreeFontMetrics(fs);
    168 #else
    169         Xfree (fs->per_char);
    170 #endif
```

Maybe there's some wrong with the bigfont support. The backtrace looks like you have not enabled it. HAve you enabled it in nxagent?

Also, this warning is suspicious:
```

warning: .dynamic section for "/usr/lib64/libfreetype.so.6" is not at the expect
ed address (wrong library or version mismatch?)
```

**Comment by mbecroft** [comment]
Thanks for the comments. I have not done anything specific to enable bigfont support (I don't even know what it is)...so I'm guessing it's not enabled by default.

Agreed the freetype thing looks suspicious, hence I have rebuilt freetype as well. Will see if there is an improvement running on latest 3.6.x HEAD with your memory leak patches merged...

**Comment by mbecroft** [comment]
Just to update this: git 3.6.x HEAD is in a broken state (see #596) so I can't yet test whether this issue is resolved in 3.6.x.

**Comment by mbecroft** [comment]
Ionic helped with a workaround for #596 so I am now testing to see if this issue is still present or resolved on HEAD as of today. Will report in a few days.

**Comment by mbecroft** [comment]
Confirm *not* fixed as of 3.6.x HEAD 3731a996dce9adf1c6f398cc1cdec2909588a7ee

Note: I rebuilt my libfreetype to be sure nothing was wrong as noted by uli42.

No improvement or change at all - problem remains exactly the same. I am struggling with how to better characterise and debug the issue! Perhaps creating a test X client which repeatably performs similar workloads to the X client applications that most frequently trigger the issue. If I can make it repeatable, then I can trace the X protocol requests and narrow in on what's triggering it.

Meanwhile, very frustrating that my web browsers generally dies about 4 times per day!! Thanks those who have tried to help me debug this... will keep working on it when I have time...

**Comment by uli42** [comment]
Hmm, are you using xfont or xfont2 (check with `pkg-config --modversion xfont2)?

**Comment by mbecroft** [comment]
```
$ pkg-config --modversion xfont2
Package xfont2 was not found in the pkg-config search path.
Perhaps you should add the directory containing `xfont2.pc'
to the PKG_CONFIG_PATH environment variable
Package 'xfont2', required by 'world', not found
$ pkg-config --modversion xfont
1.4.3
```

I'm not too sure what this library is or the difference between versions 1 and 2.

What is making us think this is a font-related issue? Is this because in the backtrace I posted it is in `XFreeFont()`?

Please bear in mind that backtrace was from when nxagent crashed with SIGBUS. That happened only once, after about 8 weeks. I have probably confused things by posting that under this ticket, as it is a separate problem.

This ticket is about the issue where X clients get randomly disconnected. When this happens, nxagent does not crash, there is no core dump. I should probably create a separate ticket for the XFreeFont() crash, but as that was on an old version (release 3.5.99.10) it may have already been fixed. So I'll only raise a ticket for that if it happens again.

We still have the unresolved issue of why certain X clients have their socket unceremoniously closed, seemingly at random, with no log/debug output from nxagent or any indication of what went wrong.

**Comment by uli42** [comment]
Ok, thanks for clarification. I still suspect your RAM or the way you compile as the reason. To verify I can try to compile a static version of nxagent. I have done some patches half a year ago to build a static version. Unfortunately the cannot be used for the current version because of the switch to autotools for libX11. So it will take some time to have this again.

However, before doing that: can you please test if my static nxagent from May is compatible to your system setup at all? You can find it here: https://drive.google.com/file/d/1ViuEGpOrWoXEOaA-szKnp1VqhKTvMvHB/view?usp=sharing

If it is runnable at your side and I can build a current one we can rule out compilation. If it shows the same problems - well, we'll see then...

**Comment by mbecroft** [comment]
Thanks. I doubt it's my RAM, as it has very specific effects and there are no other problems; also, the issue only started to happen precisely when I moved from nx 3.5 to 3.6.x prerelease.

It could be something about other libs on my system or the way it's compiled. I'll see if I can try out your static version soon and advise.

**Comment by mbecroft** [comment]
Tested the statically-linked version you referenced, no change. I'll get the following error quite often with Pale Moon (for example):

`palemoon: Fatal IO error 11 (Resource temporarily unavailable) on X server :132.`

It seems a little less frequent than before, but I think this is because the problem happens with gradually increasing frequency as the nxagent has been running over the course of weeks.

To give an idea, I've been running the static nxagent for about 48 hours. In that time, I've experienced the issue 2 or 3 times.

**Comment by mbecroft** [comment]
Minor update: this issue also occurs frequently with Audacity (version 2.1.1) and in this instance it is running on a separate host to nxagent and connecting to the X server over TCP via ssh X forwarding. I had 2 failures with the same error message as in previous comment, during a 2 hour session. It also seemed to maybe be doing the 'go slow' thing before getting dropped completely, although that's hard to quantify precisely.

So the complete list of affected X clients in my observation so far is:

- Firefox ESR 38 (have not tested more recent versions)
- Pale Moon current release
- Audacity 2.2.1
- fvwm2 2.6.5

**Comment by mbecroft** [comment]
It might also be instructive to look at software that does *not* have a problem. The following all have been rock solid and never failed in this way in now about 10 weeks of testing:

- xterm
- emacs
- PianoTeq
- Google Chrome
- Scribus
- Inkscape, The Gimp, rawtherapee (have not tested these as much)

**Comment by uli42** [comment]
We have been talking about trying to track this down and that maybe an X11 testing suite helps. The one where I could not remember the name was this: #140

**Comment by mbecroft** [comment]
Further notes: I have been doing a lot of work in Audacity and found this is one of the most reliable ways to induce the fault.

I can work on a moderately complex project in Audacity for a period of time, and faults occur in two phases (may or may not be related):

1. For the first period Audacity is running fine. This can be anywhere from 10 to 60 minutes
2. Responsiveness of *all X clients* including Audacity begins to slow down. It's possible to continue working in this phase but it becomes increasingly difficult as the responsiveness decreases.
3. Eventually, after a total of approx 30 to 120 minutes, Audacity dies with `Fatal IO error 11 (Resource temporarily unavailable) on X server :132.`

Closing and restarting Audacity (or letting it get killed by the X server) immediately fixes the responsiveness (of *all* X clients, not just Audacity).

Guessing: it starts to feel like a some kind of item leak, where as time progresses the number of items increases, perhaps leading to slowdown due to O(n) search on the leaking item. The leaky item is destroyed when the associated client (in this instance, Audacity) is disconnected from the X server, restoring normal performance.

**Comment by uli42** [comment]
Can you run xrestop in parallel when the slowdown happens?

Am 25.12.2017 21:01 schrieb "Mario Becroft" <notifications@github.com>:

**Comment by mbecroft** [comment]
Am doing so...I'm finding that by running fewer problem-causing X clients at the same time in total, and particular, having only one project open at a time in Audacity, the problems are much reduced. In the interests of getting work done I am doing that. Am monitoring with xrestop anyway to get a baseline. Will try and do more testing next few days.

**Comment by mbecroft** [comment]
A baseline xrestop from `Wed Dec 27 01:14:44 NZDT 2017`

Performance about nominal (not in go-slow state). With very few programs running. Will annotate later to explain what some of the clients are. In particular, I have closed all web browsers, except Chrome (which has been shown not to contribute to the issue).

```
xrestop - Display: localhost:0
          Monitoring 60 clients. XErrors: 48
          Pixmaps:   14817K total, Other:     559K total, All:   15377K total

res-base Wins  GCs Fnts Pxms Misc   Pxm mem  Other   Total   PID Identifier
4c00000    23   23    3   53  122    14817K      6K  14824K 20951 Guide Manager
0600000   927   26    6 2155 6821        0B    188K    188K   ?   FVWM
1200000    21  238   11   51   49        0B     18K     18K 10655 CURRENT_NOTES
6800000     6   36   14    5   22        0B     15K     15K 16387 index.php
0e00000   182   67    4    0   70        0B     11K     11K   ?   Desk
5800000     4   28    7    3  107        0B     10K     10K  3364 root@ws-mb:~
3000000     6   24    7    3  108        0B     10K     10K 31734 mb@ws-mb:~
2e00000     5   24    7    3  107        0B     10K     10K 31691 mb@ws-mb:~
6200000     4   23    7    3  107        0B     10K     10K 18245 mb@ws-mb:~
5600000     4   23    7    3  107        0B     10K     10K 23606 mb@ws-mb:~
6a00000     4   18    7    3  107        0B     10K     10K 17638 mb@ws-mb:~
5a00000     4   18    7    3  107        0B     10K     10K  9599 mb@cat:~
3e00000     4   17    8    4   16        0B      8K      8K  4851 emacs@mb-work.
5000000     3   28    5    1  106        0B      8K      8K 15976 mb@mb-work:/ho
2c00000     3   26    5    1  108        0B      8K      8K   431 mb@mb-work:/ho
4e00000     3   28    7    4   15        0B      8K      8K  2703 emacs@mb-work.
7000000     3   22    5    1  106        0B      8K      8K  7208 root@ak2:~
3c00000     3   22    5    1  106        0B      8K      8K  3009 mb@mb-work:/ho
5c00000     3   21    5    1  106        0B      8K      8K 17080 mb@ws-mb:/mnt/
6000000     3   20    5    1  106        0B      8K      8K 18238 mb@ws-mb:~
1a00000     3   18    5    1  107        0B      8K      8K 29335 mb@mb-work:~
1800000     3   18    5    1  106        0B      7K      7K 31085 mb@mb-work:~
4200000     3   16    5    1  106        0B      7K      7K 28753 mb@mb-work:~
3600000     3   15    5    1  107        0B      7K      7K   767 mb@mb-work:/ho
1600000     3   15    5    1  107        0B      7K      7K 31065 mb@mb-work:/ho
4a00000     3   15    5    1  106        0B      7K      7K 13315 mb@mb-work:/ho
5200000     3   14    5    1  106        0B      7K      7K 22793 mb@ws-mb:/mnt/
4600000     3   14    5    1  106        0B      7K      7K  5397 mb@mb-work:/ho
5400000     3   12    5    1  106        0B      7K      7K  5265 mb@mb-work:~
7200000     3   11    5    1  106        0B      7K      7K  8011 mb@mb-work:~
6e00000     3   11    5    1  106        0B      7K      7K 29559 mb@ws-mb:/mnt/
6c00000     3   11    5    1  106        0B      7K      7K 16965 mb@mb-work:~
6600000     3   11    5    1  106        0B      7K      7K  5290 mb@mb-work:~
3200000     3   11    5    1  106        0B      7K      7K 16228 mb@mb-work:/mn
2a00000     3   11    5    1  106        0B      7K      7K 32472 mb@ws-mb:~
6400000     3   10    5    1  106        0B      7K      7K 18750 mb@ws-mb:/mnt/
5e00000     3   10    5    1  106        0B      7K      7K 18186 mb@ws-mb:/mnt/
7600000     3    9    5    1  106        0B      7K      7K 12714 mb@mb-work:~
3a00000     3    9    5    1  106        0B      7K      7K 16341 mb@mb-work:/ho
3400000     3    9    5    1  106        0B      7K      7K   754 mb@mb-work:~
2800000     3    9    5    1  106        0B      7K      7K 14099 mb@ws-mb:~
1400000    29   61    1  169   63        0B      4K      4K 23856 audacity
0c00000     2    5    3    3    4        0B      3K      3K 26433 xclock
1e00000    31   30    1    0   30        0B      3K      3K   ?   Chromium clipb
3800000     6   31    1    1    6        0B      2K      2K 21413 google-chrome
4000000     4    3    1    0   21        0B      1K      1K   ?   Chromium clipb
1000000     6   15    1    2    3        0B      1K      1K   ?   imapbiff
0a00000     1    3    1    0    5        0B      1K      1K   ?   FvwmButtons
1c00000     3    1    0    0    7        0B    264B    264B 31118 google-chrome-
7800000     2    2    0    2    4        0B    192B    192B 18727 Pianoteq STAGE
2000000     0    1    0    0    2        0B     72B     72B   ?   <unknown>
7400000     1    1    0    0    0        0B     48B     48B   ?   xrestop
4400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
2400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
0800000     0    1    0    0    1        0B     48B     48B   ?   <unknown>
0200000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
4800000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2600000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2200000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
0400000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
```

**Comment by mbecroft** [comment]
I have determined the 'go-slow' and 'client abruptly disconnects' issues are not (always) related. While working with only the above-shown X clients running, Audacity with only one project open would disconnect fairly reliably after about 1 hour, with no incidences of 'go-slow'.

The two problems may still be related or could be distinct.

Also there were no obvious changes in `xrestop` leading up to the Audacity X client disconnect events.

**Comment by mbecroft** [comment]
Ok. Possibly conflating independent issues, but I have chanced upon a situation in which I can repeatably induce 'go-slow' state.

Right now I have an Audacity project, where merely starting Audacity and opening the project immediately induces 'go-slow' state. Quitting Audacity immediately restores everything to normal. That said; it seemed repeatable and consistent after about 10 times quitting and restarting Audacity with this specific project file. But then on the 11th time, everything is fine! I can't pin this thing down except to say there is some kind of correlation. Audacity seems to be the best tool yet for inducing both 'go-slow' and 'drop X client connection' issues, but I still can't tell if the two issues are related, or a  repeatable way of inducing either state--except that in some conditions simply starting up Audacity and loading a project is enough on its own to trigger 'go-slow,' but if not, then simply using it for a while *will eventually* result in either 'go-slow' or 'drop X client connection.' it is only a matter of time. I'm trying to observe whether some particular feature within Audacity brings about the issues, but so far have not been able to pin the blame to any one thing.

Another note about 'go-slow': it does not happen gradually (going from normal to slow). It happens suddenly. Everything will be normal, then for no obvious reason, everything will immediately be slow. The 'go-slow' issue affects all connected X clients, but is always resolved by terminating the 'causing' client (in this case Audacity).

`xrestop` output for the two states follows. I don't think there is much to learn from it...

In 'good' state:

```
xrestop - Display: localhost:0
          Monitoring 62 clients. XErrors: 0
          Pixmaps:       0B total, Other:     589K total, All:     589K total

res-base Wins  GCs Fnts Pxms Misc   Pxm mem  Other   Total   PID Identifier
0600000   931   26    6 2423 6797        0B    187K    187K   ?   FVWM
1200000    21  244   11   53   49        0B     18K     18K 10655 xmas2017-post-
6800000     6   37   14    5   22        0B     15K     15K 16387 index.php
0e00000   182   67    4    0   70        0B     11K     11K   ?   Desk
5800000     4   28    7    3  107        0B     10K     10K  3364 root@ws-mb:~
3000000     6   24    7    3  108        0B     10K     10K 31734 mb@ws-mb:~
2e00000     5   24    7    3  107        0B     10K     10K 31691 mb@ws-mb:~
7c00000     4   24    7    3  107        0B     10K     10K  3707 (mb@mb-work) d
7a00000     4   24    7    3  107        0B     10K     10K  3610 root@ak3:~
6200000     4   23    7    3  107        0B     10K     10K 18245 mb@ws-mb:~
5600000     4   23    7    3  107        0B     10K     10K 23606 mb@ws-mb:~
8000000     4   22    7    3  107        0B     10K     10K  3925 mb@mb-work:~
7e00000     4   21    7    3  107        0B     10K     10K  3885 mb@mb-work:~
6a00000     4   18    7    3  107        0B     10K     10K 17638 mb@ws-mb:~
5a00000     4   18    7    3  107        0B     10K     10K  9599 mb@cat:~
3e00000     4   17    8    4   16        0B      8K      8K  4851 emacs@mb-work.
5000000     3   28    5    1  106        0B      8K      8K 15976 mb@mb-work:/ho
2c00000     3   26    5    1  108        0B      8K      8K   431 mb@mb-work:/ho
4e00000     3   28    7    4   15        0B      8K      8K  2703 emacs@mb-work.
7000000     3   22    5    1  106        0B      8K      8K  7208 root@ak2:~
3c00000     3   22    5    1  106        0B      8K      8K  3009 mb@mb-work:/ho
5c00000     3   21    5    1  106        0B      8K      8K 17080 mb@ws-mb:/mnt/
6000000     3   20    5    1  106        0B      8K      8K 18238 mb@ws-mb:~
6400000     3   19    5    1  106        0B      8K      8K 18750 mb@ws-mb:/mnt/
1a00000     3   18    5    1  107        0B      8K      8K 29335 mb@mb-work:~
1800000     3   18    5    1  106        0B      7K      7K 31085 mb@mb-work:~
4200000     3   16    5    1  106        0B      7K      7K 28753 mb@mb-work:~
3600000     3   15    5    1  107        0B      7K      7K   767 mb@mb-work:/ho
1600000     3   15    5    1  107        0B      7K      7K 31065 mb@mb-work:/ho
4a00000     3   15    5    1  106        0B      7K      7K 13315 mb@mb-work:/ho
5200000     3   14    5    1  106        0B      7K      7K 22793 mb@ws-mb:/mnt/
4600000     3   14    5    1  106        0B      7K      7K  5397 mb@mb-work:/ho
5400000     3   12    5    1  106        0B      7K      7K  5265 mb@mb-work:~
7200000     3   11    5    1  106        0B      7K      7K  8011 mb@mb-work:~
6e00000     3   11    5    1  106        0B      7K      7K 29559 mb@ws-mb:/mnt/
6c00000     3   11    5    1  106        0B      7K      7K 16965 mb@mb-work:~
6600000     3   11    5    1  106        0B      7K      7K  5290 mb@mb-work:~
3200000     3   11    5    1  106        0B      7K      7K 16228 mb@mb-work:/mn
2a00000     3   11    5    1  106        0B      7K      7K 32472 mb@ws-mb:~
5e00000     3   10    5    1  106        0B      7K      7K 18186 mb@ws-mb:/mnt/
7600000     3    9    5    1  106        0B      7K      7K 12714 mb@mb-work:~
3a00000     3    9    5    1  106        0B      7K      7K 16341 mb@mb-work:/ho
3400000     3    9    5    1  106        0B      7K      7K   754 mb@mb-work:~
2800000     3    9    5    1  106        0B      7K      7K 14099 mb@ws-mb:~
0c00000     2    5    3    3    4        0B      3K      3K 26433 xclock
1e00000    31   30    1    0   30        0B      3K      3K   ?   Chromium clipb
3800000     6   31    1    1    6        0B      2K      2K 21413 google-chrome
4000000     4    3    1    0   21        0B      1K      1K   ?   Chromium clipb
1000000     6   15    1    2    3        0B      1K      1K   ?   imapbiff
0a00000     1    3    1    0    5        0B      1K      1K   ?   FvwmButtons
7800000     2    1    1    2    5        0B      1K      1K 18727 Pianoteq STAGE
1c00000     3    1    0    0    7        0B    264B    264B 31118 google-chrome-
2000000     0    1    0    0    2        0B     72B     72B   ?   <unknown>
4c00000     1    1    0    0    0        0B     48B     48B   ?   xrestop
4400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
2400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
0800000     0    1    0    0    1        0B     48B     48B   ?   <unknown>
0200000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
4800000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2600000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2200000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
0400000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
```

In 'go-slow' state:

```
xrestop - Display: localhost:0
          Monitoring 63 clients. XErrors: 94
          Pixmaps:       0B total, Other:     596K total, All:     596K total

res-base Wins  GCs Fnts Pxms Misc   Pxm mem  Other   Total   PID Identifier
0600000   945   26    6 2423 6895        0B    190K    190K   ?   FVWM
1200000    21  244   11   53   49        0B     18K     18K 10655 xmas2017-post-
6800000     6   37   14    5   22        0B     15K     15K 16387 index.php
0e00000   184   67    4    0   71        0B     11K     11K   ?   Desk
5800000     4   28    7    3  107        0B     10K     10K  3364 root@ws-mb:~
3000000     6   24    7    3  108        0B     10K     10K 31734 mb@ws-mb:~
2e00000     5   24    7    3  107        0B     10K     10K 31691 mb@ws-mb:~
7c00000     4   24    7    3  107        0B     10K     10K  3707 (mb@mb-work) d
7a00000     4   24    7    3  107        0B     10K     10K  3610 root@ak3:~
6200000     4   23    7    3  107        0B     10K     10K 18245 mb@ws-mb:~
5600000     4   23    7    3  107        0B     10K     10K 23606 mb@ws-mb:~
8000000     4   22    7    3  107        0B     10K     10K  3925 mb@mb-work:~
7e00000     4   21    7    3  107        0B     10K     10K  3885 mb@mb-work:~
6a00000     4   18    7    3  107        0B     10K     10K 17638 mb@ws-mb:~
5a00000     4   18    7    3  107        0B     10K     10K  9599 mb@cat:~
3e00000     4   17    8    4   16        0B      8K      8K  4851 emacs@mb-work.
5000000     3   28    5    1  106        0B      8K      8K 15976 mb@mb-work:/ho
2c00000     3   26    5    1  108        0B      8K      8K   431 mb@mb-work:/ho
4e00000     3   28    7    4   15        0B      8K      8K  2703 emacs@mb-work.
7000000     3   22    5    1  106        0B      8K      8K  7208 root@ak2:~
3c00000     3   22    5    1  106        0B      8K      8K  3009 mb@mb-work:/ho
5c00000     3   21    5    1  106        0B      8K      8K 17080 mb@ws-mb:/mnt/
6000000     3   20    5    1  106        0B      8K      8K 18238 mb@ws-mb:~
6400000     3   19    5    1  106        0B      8K      8K 18750 mb@ws-mb:/mnt/
1a00000     3   18    5    1  107        0B      8K      8K 29335 mb@mb-work:~
1800000     3   18    5    1  106        0B      7K      7K 31085 mb@mb-work:~
4200000     3   16    5    1  106        0B      7K      7K 28753 mb@mb-work:~
3600000     3   15    5    1  107        0B      7K      7K   767 mb@mb-work:/ho
1600000     3   15    5    1  107        0B      7K      7K 31065 mb@mb-work:/ho
4a00000     3   15    5    1  106        0B      7K      7K 13315 mb@mb-work:/ho
5200000     3   14    5    1  106        0B      7K      7K 22793 mb@ws-mb:/mnt/
4600000     3   14    5    1  106        0B      7K      7K  5397 mb@mb-work:/ho
5400000     3   12    5    1  106        0B      7K      7K  5265 mb@mb-work:~
7200000     3   11    5    1  106        0B      7K      7K  8011 mb@mb-work:~
6e00000     3   11    5    1  106        0B      7K      7K 29559 mb@ws-mb:/mnt/
6c00000     3   11    5    1  106        0B      7K      7K 16965 mb@mb-work:~
6600000     3   11    5    1  106        0B      7K      7K  5290 mb@mb-work:~
3200000     3   11    5    1  106        0B      7K      7K 16228 mb@mb-work:/mn
2a00000     3   11    5    1  106        0B      7K      7K 32472 mb@ws-mb:~
5e00000     3   10    5    1  106        0B      7K      7K 18186 mb@ws-mb:/mnt/
7600000     3    9    5    1  106        0B      7K      7K 12714 mb@mb-work:~
3a00000     3    9    5    1  106        0B      7K      7K 16341 mb@mb-work:/ho
3400000     3    9    5    1  106        0B      7K      7K   754 mb@mb-work:~
2800000     3    9    5    1  106        0B      7K      7K 14099 mb@ws-mb:~
1400000    10   51    1  169   50        0B      3K      3K 29696 audacity
0c00000     2    5    3    3    4        0B      3K      3K 26433 xclock
1e00000    31   30    1    0   30        0B      3K      3K   ?   Chromium clipb
3800000     6   31    1    1    6        0B      2K      2K 21413 google-chrome
4000000     4    3    1    0   21        0B      1K      1K   ?   Chromium clipb
1000000     6   15    1    2    3        0B      1K      1K   ?   imapbiff
0a00000     1    3    1    0    5        0B      1K      1K   ?   FvwmButtons
7800000     2    1    1    2    5        0B      1K      1K 18727 Pianoteq STAGE
1c00000     3    1    0    0    7        0B    264B    264B 31118 google-chrome-
2000000     0    1    0    0    2        0B     72B     72B   ?   <unknown>
7400000     1    1    0    0    0        0B     48B     48B   ?   xrestop
4400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
2400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
0800000     0    1    0    0    1        0B     48B     48B   ?   <unknown>
0200000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
4800000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2600000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2200000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
0400000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
```

**Comment by mbecroft** [comment]
For reference, the 'drops X client connection' issue manifests as follows in Audacity:

```
[mb@ws-mb xmas2017-audio]$ audacity
<some debug output from Audacity, ALSA etc.>
<everything is fine for some time, then:>
audacity: Fatal IO error 11 (Resource temporarily unavailable) on X server localhost:11.0.
audacity: Fatal IO error 11 (Resource temporarily unavailable) on X server localhost:11.0.
[mb@ws-mb xmas2017-audio]$ 
```

The error message always appears twice in the case of Audacity, whereas with other affected clients such as Pale Moon, the message appears only once. I'm guessing Audacity has multiple threads accessing the X server perhaps?

In this case Audacity is connecting via ssh X forwarding. The issue happens identically with clients running on the same host as `nxagent` as well, but then, of course, the display number in the error message emitted by the affected X client is is the nxagent's display number.

It has just occured to me though, I wonder whether Audacity is so good at triggering the issue because of something inherent to Audacity, or because it is running over ssh forwarding.

I have another X client that is always running (emacs) which runs on another host and connects directly via TCP (not via ssh). This so far has never been the 'cause' of go-slow nor affected by the 'drop X connection' issue.

**Comment by mbecroft** [comment]
Here is another picture of `xrestop` with Pale Moon (*and* Audacity) running.

Starting up Pale Moon while Audacity is running seems to be absolutely killer. Things seemed ok for a few minutes, then `nxagent` became so unresponsive that it would take several minutes just to move focus to another window, or create a new xterm instance. `nxagent` would become completely unresponsive to X client requests for up to a minute at a time.

It's a bit harder to run `xrestop` while things have become unresponsive, but it was not evident that anything changed in `xrestop` when the problem started.

Killing Pale Moon instantly brought things back to normal.

As always, this is all rather hard to pin down, since after restarting Pale Moon again (with Audacity still running,) everything has been fine and responding normally for at least 10 minutes!

FWIW, during this whole time, my nx connection to the nxagent remained up.

The 0B Pxm value for all clients seems a bit suspicious. As is the first line: `Display: localhost:0`. It is definitely showing output for the `nxagent` display, in this instance :132.0:

```
mb@mb-work ~ $ echo $DISPLAY
:132
```

```
xrestop - Display: localhost:0
          Monitoring 65 clients. XErrors: 34
          Pixmaps:       0B total, Other:     734K total, All:     734K total

res-base Wins  GCs Fnts Pxms Misc   Pxm mem  Other   Total   PID Identifier
0600000  1365   26    6 2617 9780        0B    267K    267K   ?   FVWM
8400000    74   33    1 1284 2158        0B     54K     54K  2891 Pale Moon
6800000     6   37   14    5   22        0B     15K     15K 16387 index.php
1200000    21  219    8   49   49        0B     14K     14K 10655 xmas2017-post-
0e00000   244   67    4    0  101        0B     13K     13K   ?   Desk
5800000     4   28    7    3  107        0B     10K     10K  3364 root@ws-mb:~
3000000     6   24    7    3  108        0B     10K     10K 31734 mb@ws-mb:~
2e00000     5   24    7    3  107        0B     10K     10K 31691 mb@ws-mb:~
7c00000     4   24    7    3  107        0B     10K     10K  3707 mb@mb-work:~
7a00000     4   24    7    3  107        0B     10K     10K  3610 root@ak3:~
6200000     4   23    7    3  107        0B     10K     10K 18245 mb@ws-mb:~
5600000     4   23    7    3  107        0B     10K     10K 23606 mb@ws-mb:~
8000000     4   22    7    3  107        0B     10K     10K  3925 mb@mb-work:~
7e00000     4   21    7    3  107        0B     10K     10K  3885 mb@mb-work:~
6a00000     4   18    7    3  107        0B     10K     10K 17638 mb@ws-mb:~
5a00000     4   18    7    3  107        0B     10K     10K  9599 mb@cat:~
3e00000     4   17    8    4   16        0B      8K      8K  4851 emacs@mb-work.
5000000     3   28    5    1  106        0B      8K      8K 15976 mb@mb-work:/ho
2c00000     3   26    5    1  108        0B      8K      8K   431 mb@mb-work:/ho
1800000     3   24    5    1  107        0B      8K      8K 31085 (mb@mb-work) d
4e00000     3   28    7    4   15        0B      8K      8K  2703 emacs@mb-work.
7000000     3   22    5    1  106        0B      8K      8K  7208 root@ak2:~
3c00000     3   22    5    1  106        0B      8K      8K  3009 mb@mb-work:/ho
5c00000     3   21    5    1  106        0B      8K      8K 17080 mb@ws-mb:/mnt/
6000000     3   20    5    1  106        0B      8K      8K 18238 mb@ws-mb:~
6400000     3   19    5    1  106        0B      8K      8K 18750 mb@ws-mb:/mnt/
1a00000     3   18    5    1  107        0B      8K      8K 29335 mb@mb-work:~
7400000     3   17    5    1  106        0B      7K      7K 21710 mb@mb-work:/ho
1600000     3   16    5    1  107        0B      7K      7K 31065 mb@mb-work:/ho
4200000     3   16    5    1  106        0B      7K      7K 28753 mb@mb-work:~
3600000     3   15    5    1  107        0B      7K      7K   767 mb@mb-work:/ho
4a00000     3   15    5    1  106        0B      7K      7K 13315 mb@mb-work:/ho
5200000     3   14    5    1  106        0B      7K      7K 22793 mb@ws-mb:/mnt/
4600000     3   14    5    1  106        0B      7K      7K  5397 mb@mb-work:/ho
5400000     3   12    5    1  106        0B      7K      7K  5265 mb@mb-work:~
7200000     3   11    5    1  106        0B      7K      7K  8011 mb@mb-work:~
6e00000     3   11    5    1  106        0B      7K      7K 29559 mb@ws-mb:/mnt/
6c00000     3   11    5    1  106        0B      7K      7K 16965 mb@mb-work:~
6600000     3   11    5    1  106        0B      7K      7K  5290 mb@mb-work:~
3200000     3   11    5    1  106        0B      7K      7K 16228 mb@mb-work:/mn
2a00000     3   11    5    1  106        0B      7K      7K 32472 mb@ws-mb:~
5e00000     3   10    5    1  106        0B      7K      7K 18186 mb@ws-mb:/mnt/
7600000     3    9    5    1  106        0B      7K      7K 12714 mb@mb-work:~
3a00000     3    9    5    1  106        0B      7K      7K 16341 mb@mb-work:/ho
3400000     3    9    5    1  106        0B      7K      7K   754 mb@mb-work:~
2800000     3    9    5    1  106        0B      7K      7K 14099 mb@ws-mb:~
1400000    23   56    1  169   59        0B      4K      4K 29719 audacity
0c00000     2    5    3    3    4        0B      3K      3K 26433 xclock
1e00000    30   29    1    0   30        0B      3K      3K   ?   Chromium clipb
3800000     6   31    1    1    6        0B      2K      2K 21413 google-chrome
4000000     4    3    1    0   21        0B      1K      1K   ?   Chromium clipb
1000000     6   15    1    2    3        0B      1K      1K   ?   imapbiff
0a00000     1    3    1    0    5        0B      1K      1K   ?   FvwmButtons
7800000     2    1    1    2    5        0B      1K      1K 18727 Pianoteq STAGE
1c00000     3    1    0    0    7        0B    264B    264B 31118 google-chrome-
2000000     0    1    0    0    2        0B     72B     72B   ?   <unknown>
4c00000     1    1    0    0    0        0B     48B     48B   ?   xrestop
4400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
2400000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
0800000     0    1    0    0    1        0B     48B     48B   ?   <unknown>
0200000     1    1    0    0    0        0B     48B     48B   ?   <unknown>
4800000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2600000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
2200000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
0400000     0    1    0    0    0        0B     24B     24B   ?   <unknown>
```

**Comment by uli42** [comment]
Hm, so it is probably not related to resources. Just to make sure we are collecting all information about the case:
- are you running xrestop inside the session (which would mean the update of xrestop's output would create further activity in the nxagent) or are you attaching the xrestop from the outside (DISPLAY=:something xrestop)?

Some further ideas to approach the problem (unfortunately I cannot test this here because I have only seen this on the system in my office and I am on vacation currently):
- You say with 3.5.0 this does not occur. You could try to run a 3.5.0 session but replace the libNX_X11 by the 3.5.99.x version. We then could deduce if the problem is connected to the libNX_X11 or rather the nxagent. Same for libXcomp (and libXcompshad although that should not matter here).
- run nxagent with `-rrxinerama` (to ensure that my xinerama code is out of the game)
- compile nxagent with DEBUG, WARNING and TEST output enabled in hw/nxagent/Events.c (line 115/116)
- run nxagent without nxcomp: xauth add :123 MIT-MAGIC-COOKIE <generated cookie>; nxagent :123 and then run your desktop environment in that nxagent. If the problem disappears we should have a look at nxcomp.
- I have a further idea reagrding serials but I need to do some research first.

**Comment by uli42** [comment]
Here details to my further idea: Search for `LostSequence`  in `nx-X11/programs/Xserver/hw/nxagent/compext/Compext.c` and remove the `#ifdef TEST` block in the found functions. Watch if any messages regarding lost sequence numbers can be connected to the slowdown to the IO errors. 
(You can also simply change `#undef TEST` to `#define TEST` right at the top of that file but it will produce many output lines...)

(and, BTW, use the current HEAD, Ionic has fixed some more issues in the logging code and I have fixed many memory leaks)

**Comment by uli42** [comment]
I have not fully read into this, but it certainly is interesting: https://stackoverflow.com/questions/23871516/xreply-terminates-app-with-xioerror 
-> does audacity use libxcb? Does palemoon?

Another fix that needs investigation: https://lists.x.org/archives/xorg-devel/2015-September/047338.html _fix for Xlib 32-bit request number issues_

**Comment by mbecroft** [comment]
Thanks a lot, uli42, for all your research and suggestions. Agreed, it does not seem related to resources, at least not per se as reported by `xrestop`.

The question you found on Stack Overflow may be relevant as their symptoms are identical to mine.

Both Audacity and Pale Moon are indeed linked with `libxcb`. However, both are using `libxcb` 1.11, much newer than the version with the issue, so one would expect it has been long since fixed.

Later, I will try building one of the trivial (1-liner) test clients from that post to see if this reproduces the 'dropped X connection' problem. If I can reproduce it at will, this would be a breakthrough. I could also leave gdb attached to, say, Audacity to see whether the X I/O error is emitted from the same location as in the SO issue. They did mention the program had to be compiled with `gcc -m32`. All my libraries and clients are targetted at `amd64`.

I'm still puzzled though, as these same applications were working fine on earlier `nx-libs` 3.5, using the same `libxcb`. Perhaps the issue is not directly in my `libxcb` but a similar problem elsewhere in the `nx-libs` codebase.

You've given me plenty of other ways to test and narrow down the issue. I'll attempt to try those out soon... it's tough as the problems are intermittent. Will update this ticket as soon as I have learned something.

**Comment by mbecroft** [comment]
So I have started running the `XDrawLine` test that was "the original code snippet from the submitted bug report which was used to reproduce the issue" in the SO post. I am running both normal 64-bit and a version built with `gcc -m32` in parallel. OP said it could take about 15 hours for them to see the issue with their application. Let's see what happens...

**Comment by mbecroft** [comment]
After approx 24 hours, the XDrawLine test compiled with `gcc -m32` failed with a similar error:

```
$ ./nx-test2-32bit
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":132"
      after 61356 requests (61356 known processed) with 25 events remaining.
$ 
```

The version built without `-m32` (targetting `x86_64` architecture) has not *yet* failed.

Not having time to investigate and understand this more deeply, I shall now conduct the following further tests:

- Add counting and reporting of number of XDrawLine calls (surely it must have done more than the reported 61K drawing operations in 24 hours!)
- Run the same test programs displaying to a more standard nVidia Xorg server and compare the behaviour.

**Comment by uli42** [comment]
Hmm, I am wondering how audacity manages to reach that situation much
quicker!?

Am 31.12.2017 8:00 vorm. schrieb "Mario Becroft" <notifications@github.com>:

**Comment by mbecroft** [comment]
Yes, clearly there is more to this than we currently understand. I'm sure we will get to the bottom of it...I can't rule out an issue with the other libs on my system (libxcb especially) even though the version I have certainly no longer has the bug originally reported on SO. It was mentioned that even just calling XNoOp() results in the same issue. This might get there faster, who knows...will have more info soon. Ran out of time to do more testing today.

It's bang on midnight here and I can hear all the fireworks. Am happily wrapped up in bed and hopefully going to sleep shortly myself, however. Happy New Year!

**Comment by mbecroft** [comment]
Turns out libxcb 1.11 is much older than I thought--at least 2 years old. And could well still have the bug that was reported.

So, the test code changed to do XNoOp() gets through operations quite fast...and when the test code is compiled for 32-bit x86 (gcc -m32) it exhibits the disconnect after just a couple of hours. I added counting the draw commands and printing the sequence from `_XPrivDisplay->request`. Turns out it happens when the 32-bit sequence wraps around from 4 billion to 0. The fault then occurs about 60K requests after the wrap-around, hence the ~60K number in the message (in reality there have been 4 billion + ~60K requests).

Confirmed that in libxcb 1.12 (vs 1.11) this is fixed for 32-bit clients.

Test program never exhibits the issue on 64-bit (x86_64); the sequence just keeps climbing beyond 4 billion.

As 64-bit test program does not exhibit the issue, it's unclear whether this is the same or a different issue to the one causing disconnects of 64-bit clients from nxagent. But perhaps depending on certain interactions of the client and nxagent, a related issue occurs. Or, it is unrelated and just happens to result in similar symptoms.

I am now running Pale Moon, Audacity and fvwm2, which most frequently exhibit the fault, to see if these are now fixed with libxcb 1.12.

So far, no crash after about 5 hours...

**Comment by mbecroft** [comment]
Test program:

```C
#include <X11/Xlib.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
  Display *d;
  int s;
  Window w;
  XEvent e;
  unsigned long c;

  /* open connection with the server */
  d=XOpenDisplay(NULL);

  if(!d) {
    printf("Cannot open display\n");
    exit(1);
  }

  s=DefaultScreen(d);

  w=XCreateSimpleWindow(d, RootWindow(d, s), 10, 10, 100, 100, 1,
                        BlackPixel(d, s), WhitePixel(d, s));

  Atom delWindow = XInternAtom( d, "WM_DELETE_WINDOW", 0 );
  XSetWMProtocols(d , w, &delWindow, 1);

  /* XSelectInput(d, w, ExposureMask | KeyPressMask); */

  XMapWindow(d, w);

  for(c = 0;;c++) {
    if (c % 1000000 == 0) {
      printf(
             "Operations: %lu,    sequence %ld\n",
             c,
             ((_XPrivDisplay)d)->request
             );
    }
    /* XDrawLine(d, w, DefaultGC(d, s), 10, 60, 180, 20); */
    XNoOp(d);
    XFlush(d);
  }

  XDestroyWindow(d, w);
  XCloseDisplay(d);

  return 0;
}
```

**Comment by uli42** [comment]
I have compiled a little test program with 32bit (`gcc -m32 x11test2.c  -lX11 -o x11test`):
```
#include <X11/Xlib.h>
void main(void) {
    Display *d = XOpenDisplay(NULL);
    if (d)
     for(;;)
        XNoOp(d);
}
```

I ran it against latest nxagent (without nxcomp, just nxagent as xnest replacement): `nxagent :55`, then `DISPLAY=:55 time ./x11test`. It ran for approximately 3 hours and then stopped:

```
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":55"
      after 62982 requests (62982 known processed) with 0 events remaining.
Command exited with non-zero status 1
84.40user 14.46system 2:49:12elapsed 0%CPU (0avgtext+0avgdata 1968maxresident)k
2096inputs+0outputs (15major+96minor)pagefaults 0swaps
```

The nxagent only reported 
```
Program received signal SIGPIPE, Broken pipe.
```

A previously attached gdb showed the same value as the client's sequence (62982):
```
(gdb) print ((ClientPtr) 0xdf90a0)->sequence
$12 = 62982
(gdb) print *(ClientPtr) 0xdf90a0
$13 = {index = 14717872, clientAsMask = 0, requestBuffer = 0xdf94a0, osPrivate = 0x0, swapped = 0, local = 1, pSwapReplyFunc = 0xdf8550, errorValue = 111, sequence = 62982, closeDownMode = 0, clientGone = 1, noClientException = -1, 
  saveSet = 0x0, numSaved = 0, screenPrivate = {0xdf87c0, 0xdf8740, 0xdf8b10, 0xdf8b90, 0xdf8d10, 0xdf8d90, 0xdf8f30, 0xdf8fb0, 0x2d6344b290b36347, 0x29e1, 0x7ffff54e7b78 <main_arena+88>, 0xdf8e90, 0x0, 0x0, 0x6b635896acb5902c, 
    0x9f68d1dad49626d4}, requestVector = 0xb2fee0 <ProcVector>, req_len = 1, big_requests = 1, priority = 0, clientState = ClientStateGone, devPrivates = 0xdf91e8, xkbClientFlags = 32768, mapNotifyMask = 0, 
  newKeyboardNotifyMask = 13409, vMajor = 1, vMinor = 0, minKC = 8 '\b', maxKC = 255 '\377', replyBytesRemaining = 0, authId = 289, trustLevel = 0, CheckAccess = 0x0, fontResFunc = 0x0, smart_priority = 0, smart_start_tick = 0, 
  smart_stop_tick = 0, smart_check_tick = 0, clientIds = 0x0}
```

The system libraries on my Ubuntu 16.04 are:
```
$ dpkg -l | grep libx11
ii  libx11-6:amd64                                                 2:1.6.3-1ubuntu2                                         amd64        X11 client-side library
ii  libx11-6:i386                                                  2:1.6.3-1ubuntu2                                         i386         X11 client-side library
ii  libx11-6-dbg:amd64                                             2:1.6.3-1ubuntu2                                         amd64        X11 client-side library (debug package)
ii  libx11-data                                                    2:1.6.3-1ubuntu2                                         all          X11 client-side library
ii  libx11-dev:amd64                                               2:1.6.3-1ubuntu2                                         amd64        X11 client-side library (development headers)
ii  libx11-dev:i386                                                2:1.6.3-1ubuntu2                                         i386         X11 client-side library (development headers)
ii  libx11-doc                                                     2:1.6.3-1ubuntu2                                         all          X11 client-side library (development documentation)
ii  libx11-protocol-perl                                           0.56-7                                                   all          Perl module for the X Window System Protocol, version 11
ii  libx11-xcb-dev:amd64                                           2:1.6.3-1ubuntu2                                         amd64        Xlib/XCB interface library (development headers)
ii  libx11-xcb1:amd64                                              2:1.6.3-1ubuntu2                                         amd64        Xlib/XCB interface library
ii  libx11-xcb1:i386                                               2:1.6.3-1ubuntu2                                         i386         Xlib/XCB interface library
$ dpkg -l | grep libxcb
ii  libxcb-composite0:amd64                                        1.11.1-1ubuntu1                                          amd64        X C Binding, composite extension
ii  libxcb-cursor0:amd64                                           0.1.1-3ubuntu1                                           amd64        utility libraries for X C Binding -- cursor
ii  libxcb-damage0:amd64                                           1.11.1-1ubuntu1                                          amd64        X C Binding, damage extension
ii  libxcb-dpms0:amd64                                             1.11.1-1ubuntu1                                          amd64        X C Binding, dpms extension
ii  libxcb-dri2-0:amd64                                            1.11.1-1ubuntu1                                          amd64        X C Binding, dri2 extension
ii  libxcb-dri2-0:i386                                             1.11.1-1ubuntu1                                          i386         X C Binding, dri2 extension
ii  libxcb-dri2-0-dev:amd64                                        1.11.1-1ubuntu1                                          amd64        X C Binding, dri2 extension, development files
ii  libxcb-dri3-0:amd64                                            1.11.1-1ubuntu1                                          amd64        X C Binding, dri3 extension
ii  libxcb-dri3-0:i386                                             1.11.1-1ubuntu1                                          i386         X C Binding, dri3 extension
ii  libxcb-dri3-dev:amd64                                          1.11.1-1ubuntu1                                          amd64        X C Binding, dri3 extension, development files
ii  libxcb-glx0:amd64                                              1.11.1-1ubuntu1                                          amd64        X C Binding, glx extension
ii  libxcb-glx0:i386                                               1.11.1-1ubuntu1                                          i386         X C Binding, glx extension
ii  libxcb-glx0-dev:amd64                                          1.11.1-1ubuntu1                                          amd64        X C Binding, glx extension, development files
ii  libxcb-icccm4:amd64                                            0.4.1-1ubuntu1                                           amd64        utility libraries for X C Binding -- icccm
ii  libxcb-image0:amd64                                            0.4.0-1build1                                            amd64        utility libraries for X C Binding -- image
ii  libxcb-keysyms1:amd64                                          0.4.0-1                                                  amd64        utility libraries for X C Binding -- keysyms
ii  libxcb-present-dev:amd64                                       1.11.1-1ubuntu1                                          amd64        X C Binding, present extension, development files
ii  libxcb-present0:amd64                                          1.11.1-1ubuntu1                                          amd64        X C Binding, present extension
ii  libxcb-present0:i386                                           1.11.1-1ubuntu1                                          i386         X C Binding, present extension
ii  libxcb-randr0:amd64                                            1.11.1-1ubuntu1                                          amd64        X C Binding, randr extension
ii  libxcb-randr0-dev:amd64                                        1.11.1-1ubuntu1                                          amd64        X C Binding, randr extension, development files
ii  libxcb-record0:amd64                                           1.11.1-1ubuntu1                                          amd64        X C Binding, record extension
ii  libxcb-render-util0:amd64                                      0.3.9-1                                                  amd64        utility libraries for X C Binding -- render-util
ii  libxcb-render0:amd64                                           1.11.1-1ubuntu1                                          amd64        X C Binding, render extension
ii  libxcb-render0:i386                                            1.11.1-1ubuntu1                                          i386         X C Binding, render extension
ii  libxcb-render0-dev:amd64                                       1.11.1-1ubuntu1                                          amd64        X C Binding, render extension, development files
ii  libxcb-shape0:amd64                                            1.11.1-1ubuntu1                                          amd64        X C Binding, shape extension
ii  libxcb-shape0-dev:amd64                                        1.11.1-1ubuntu1                                          amd64        X C Binding, shape extension, development files
ii  libxcb-shm0:amd64                                              1.11.1-1ubuntu1                                          amd64        X C Binding, shm extension
ii  libxcb-shm0:i386                                               1.11.1-1ubuntu1                                          i386         X C Binding, shm extension
ii  libxcb-shm0-dev:amd64                                          1.11.1-1ubuntu1                                          amd64        X C Binding, shm extension, development files
ii  libxcb-sync-dev:amd64                                          1.11.1-1ubuntu1                                          amd64        X C Binding, sync extension, development files
ii  libxcb-sync1:amd64                                             1.11.1-1ubuntu1                                          amd64        X C Binding, sync extension
ii  libxcb-sync1:i386                                              1.11.1-1ubuntu1                                          i386         X C Binding, sync extension
ii  libxcb-util0:amd64                                             0.3.8-2ubuntu1                                           amd64        utility libraries for X C Binding -- atom, aux and event
ii  libxcb-util1:amd64                                             0.4.0-0ubuntu3                                           amd64        utility libraries for X C Binding -- atom, aux and event
ii  libxcb-xf86dri0:amd64                                          1.11.1-1ubuntu1                                          amd64        X C Binding, xf86dri extension
ii  libxcb-xfixes0:amd64                                           1.11.1-1ubuntu1                                          amd64        X C Binding, xfixes extension
ii  libxcb-xfixes0-dev:amd64                                       1.11.1-1ubuntu1                                          amd64        X C Binding, xfixes extension, development files
ii  libxcb-xkb1:amd64                                              1.11.1-1ubuntu1                                          amd64        X C Binding, XKEYBOARD extension
ii  libxcb-xv0:amd64                                               1.11.1-1ubuntu1                                          amd64        X C Binding, xv extension
ii  libxcb-xv0-dev:amd64                                           1.11.1-1ubuntu1                                          amd64        X C Binding, xv extension, development files
ii  libxcb-xvmc0:amd64                                             1.11.1-1ubuntu1                                          amd64        X C Binding, xvmc extension
ii  libxcb-xvmc0-dev:amd64                                         1.11.1-1ubuntu1                                          amd64        X C Binding, xvmc extension, development files
ii  libxcb1:amd64                                                  1.11.1-1ubuntu1                                          amd64        X C Binding
ii  libxcb1:i386                                                   1.11.1-1ubuntu1                                          i386         X C Binding
ii  libxcb1-dev:amd64                                              1.11.1-1ubuntu1                                          amd64        X C Binding, development files
ii  libxcb1-dev:i386                                               1.11.1-1ubuntu1                                          i386         X C Binding, development files
```

**Comment by mbecroft** [comment]
You have confirmed the known bug in `libxcb` 1.11 when built for x86 (32-bit). This happens when connecting to any X server (or, at least, any Xorg-based server), not just `nxagent`. It is fixed in `libxcb` 1.12. Fortunately RHEL now has `libxcb` 1.12.

I am not convinced that this is the cause of my issues reported in this ticket, but it looks increasingly like there is more than one cause. I am now focusing on building the latest nx-libs and testing again with all clients running on `libxcb` 1.12.

**Comment by uli42** [comment]
Yes, I just wanted to record the state of the nxagent when that happens
(for reference). I am now doing the same with 1.12.

Am 07.01.2018 21:37 schrieb "Mario Becroft" <notifications@github.com>:

**Comment by uli42** [comment]
Now, this is interesting: Using libxcb-1.12 does not change anything! Same behaviour, same sequence number:

```
$ LD_LIBRARY_PATH=/home/uli/Downloads/libxcb-1.12/src/.libs/ ldd x11test2
        linux-gate.so.1 =>  (0xf779e000)
        libX11.so.6 => /usr/lib/i386-linux-gnu/libX11.so.6 (0xf760c000)
        libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xf7456000)
        libxcb.so.1 => /home/uli/Downloads/libxcb-1.12/src/.libs/libxcb.so.1 (0xf7423000)
        libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xf741e000)
        /lib/ld-linux.so.2 (0x56652000)
        libXau.so.6 => /usr/lib/i386-linux-gnu/libXau.so.6 (0xf741a000)
        libXdmcp.so.6 => /usr/lib/i386-linux-gnu/libXdmcp.so.6 (0xf7413000)
```

```
(gdb) print *((ClientPtr) 0xe170a0)
$19 = {index = 14650000, clientAsMask = 0, requestBuffer = 0x7ffff54e7b78 <main_arena+88>, osPrivate = 0x0, swapped = 0, local = 0, pSwapReplyFunc = 0x0, errorValue = 111, sequence = 62982, closeDownMode = 0, clientGone = 1, 
  noClientException = -1, saveSet = 0x0, numSaved = 0, screenPrivate = {0x0 <repeats 16 times>}, requestVector = 0xb2fee0 <ProcVector>, req_len = 1, big_requests = 1, priority = 0, clientState = ClientStateGone, devPrivates = 0xe171e8, 
  xkbClientFlags = 32768, mapNotifyMask = 0, newKeyboardNotifyMask = 0, vMajor = 1, vMinor = 0, minKC = 8 '\b', maxKC = 255 '\377', replyBytesRemaining = 0, authId = 289, trustLevel = 0, CheckAccess = 0x0, fontResFunc = 0x0, 
  smart_priority = 0, smart_start_tick = 0, smart_stop_tick = 0, smart_check_tick = 0, clientIds = 0x0}
```

Seems like there's also a update of libX11 required.

**Comment by mbecroft** [comment]
That may well be so. I upgraded both libX11 and libxcb at the same time, which did resolve the problem.

**Comment by uli42** [comment]
https://cgit.freedesktop.org/xorg/lib/libX11/commit/?id=a72d2d06c002b644b7040a0a9936c8525e092ba8

Which was integrated between libX11 1.6.3 and 1.6.4. And my libx11 is at 2:1.6.3-1ubuntu2...

**Comment by uli42** [comment]
I have just experienced the crash of plasma-desktop on my Centos7.4 machine. And this time it was running through xtrace. The last 10 lines of the logfile are these:
```
000:<:8509b9: 36: RENDER-Request(144,8): Composite op=Src(0x01) src=0x01c005a2 mask=None(0x00000000) dst=0x01dffffd xSrc=0 ySrc=0 xMask=218 yMask=25 xDst=0 yDst=0 width=218 height=25
000:<:8509ba: 16: Request(55): CreateGC cid=0x01dffffe drawable=0x01dffffc values={}
000:<:8509bb: 16: Request(55): CreateGC cid=0x01dfffff drawable=0x01dffffc values={}
000:<:8509bc: 16: Request(98): QueryExtension name='XC-MISC'
000:>:09bc:32: Reply to QueryExtension: present=false(0x00) major-opcode=0 first-event=0 first-error=0
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b741 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2201 root-y=1192 event-x=2201 event-y=27 state=0 same-screen=true(0x01)
000:>:8509bc: Event (generated) ClientMessage(33) format=0x20 window=0x01c002fa type=0x1b1("_NET_SYSTEM_TRAY_OPCODE") data=0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x8b,0x00,0x00,0x03,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00;
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b749 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2201 root-y=1191 event-x=2201 event-y=26 state=0 same-screen=true(0x01)
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b751 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2201 root-y=1190 event-x=2201 event-y=25 state=0 same-screen=true(0x01)
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b759 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2202 root-y=1190 event-x=2202 event-y=25 state=0 same-screen=true(0x01)
```

So _it looks like_ it might have something to do with the tray or with the XC-MISC extension (which I have on my local nvidia x server, but not in nxagent)

**Comment by uli42** [comment]
The apperance of XC-MISC in the output indicates that we are running out of ressource IDs here (see https://www.x.org/releases/X11R7.6/doc/xcmiscproto/xc-misc.html for the sole purpose of this extension). 

If that really is the problem then we should probably backport the XC-MISC extension. However, this would not explain why it does not happen with 3.5.0

**Comment by uli42** [comment]
Ah, even more material:

When plasma-desktop extablishes the connection to the X server it gets a range of resource IDs:

```
000:>: Success, version is 11:0 vendor='The X.Org Foundation' release=70000000 resource-id=0x01c00000 resource-mask=0x001fffff motion-buffer-size=256 max-request-len=65535 image-byte-order=LSBFirst(0x00) bitmap-bit-order=LeastSignificant(0x00) scanline-unit=32 scanline-pad=32 min-keycode=0x08 max-keycode=0x00 pixmap-formats={depth=1 bits/pixel=1 scanline-pad=32},{depth=4 bits/pixel=8 scanline-pad=32},{depth=8 bits/pixel=8 scanline-pad=32},{depth=15 bits/pixel=16 scanline-pad=32},{depth=16 bits/pixel=16 scanline-pad=32},{depth=24 bits/pixel=32 scanline-pad=32},{depth=32 bits/pixel=32 scanline-pad=32}; roots={root=0x000000f0 default-colormap=0x000000d6 white-pixel=0x00ffffff black-pixel=0x00000000 input-mask=KeyPress,KeyRelease,EnterWindow,LeaveWindow,KeymapState,Exposure,StructureNotify,SubstructureNotify,SubstructureRedirect,FocusChange,PropertyChange,ColormapChange width[pixel]=3840 height[pixel]=1200 width[mm]=1016 height[mm]=318 min-installed-maps=1 max-installed-maps=1 root=0x000000cd backing-stores=Always(0x02) save-unders=true(0x01) root-depth=24 allowed depths={depth=24 visuals={id=0x000000cd class=TrueColor(0x04) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000ce class=TrueColor(0x04) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000cf class=TrueColor(0x04) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000d0 class=TrueColor(0x04) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000d1 class=DirectColor(0x05) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000d2 class=DirectColor(0x05) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000d3 class=DirectColor(0x05) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff},{id=0x000000d4 class=DirectColor(0x05) bits/rgb-value=11 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff};},{depth=1 visuals=;},{depth=4 visuals=;},{depth=8 visuals=;},{depth=15 visuals=;},{depth=16 visuals=;},{depth=32 visuals=;};};
```

The range here is `resource-id=0x01c00000 resource-mask=0x001fffff`, probably (need to verify) this means the range is 0x01c00000 to 0x01dfffff. In my xtrace output I see that plasma-desktop is working with resource ids close to the end of that range (see `dst=`, `drawable=` and `cid=` in the following output):

```
00:<:8509b9: 36: RENDER-Request(144,8): Composite op=Src(0x01) src=0x01c005a2 mask=None(0x00000000) dst=0x01dffffd xSrc=0 ySrc=0 xMask=218 yMask=25 xDst=0 yDst=0 width=218 height=25
000:<:8509ba: 16: Request(55): CreateGC cid=0x01dffffe drawable=0x01dffffc values={}
000:<:8509bb: 16: Request(55): CreateGC cid=0x01dfffff drawable=0x01dffffc values={}
000:<:8509bc: 16: Request(98): QueryExtension name='XC-MISC'
000:>:09bc:32: Reply to QueryExtension: present=false(0x00) major-opcode=0 first-event=0 first-error=0
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b741 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2201 root-y=1192 event-x=2201 event-y=27 state=0 same-screen=true(0x01)
000:>:8509bc: Event (generated) ClientMessage(33) format=0x20 window=0x01c002fa type=0x1b1("_NET_SYSTEM_TRAY_OPCODE") data=0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x8b,0x00,0x00,0x03,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00;
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b749 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2201 root-y=1191 event-x=2201 event-y=26 state=0 same-screen=true(0x01)
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b751 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2201 root-y=1190 event-x=2201 event-y=25 state=0 same-screen=true(0x01)
000:>:8509bc: Event MotionNotify(6) detail=Normal(0x00) time=0xd532b759 root=0x000000f0 event=0x01c003ce child=None(0x00000000) root-x=2202 root-y=1190 event-x=2202 event-y=25 state=0 same-screen=true(0x01)
``` 

For me this looks like the client is exhausting its range and libX11 (or merely libxcb) is trying to use XC-MISC to resolve that situation. But as the extension is missing it cannot. And then the client gets an error from libX11 (all this is subject to verification).

**Comment by uli42** [comment]
We can probably verify my theories above my writing a test client that rapidly allocates (and maybe frees) resources. Or by lowering the id range the client gets (should be defined by RESOURCE_AND_CLIENT_COUNT and/or  RESOURCE_CLIENT_BITS).

**Comment by mbecroft** [comment]
This is a reasonable hypothesis, as the X server gives the client some resource mask initially, but if resource identifiers run out the client can only crash, and the number of available resource identifiers initially provided is somewhat arbitrary. This corresponds with my observation that processes with larger numbers of windows and tabs open, for example, seem more susceptible to the problem. On the other hand, xrestop did not indicate an excessive number of resources--but perhaps xrestop does not count all resources, or resource identifiers cannot always be reused leading to a problem even if it would appear there should be sufficient.

Your xtrace output is extremely helpful in narrowing this down!

Why it did not happen before? Perhaps the older version *did* have XC-MISC, or handed out much larger resource pools in the first place, or the pre-xcb libX11 in use by me or others at that time might have a different resource pool management strategy.

**Comment by uli42** [comment]
Well, the whole point of XC-MISC is that the _XIDs_ are exhausted, as XIDs are not re-used after having been freed. This does not mean that the resources are allocated at all.

Interestingly the old libX11code without XCB had an explicit error about resources being exhausted but I have not checked the current state yet.

**Comment by mbecroft** [comment]
Either way it's quite clear that your specific crash in this instance was resource exhaustion...why that situation arose is less clear...if I understand correctly, nx-libs 3.6.x are now rebased off upstream Xorg, so I think it should be trivial to enable the XC-MISC extension? However, other parts of nx libraries may well make assumptions about resource assignment that have to be taken into account.

**Comment by uli42** [comment]
No, upstream Xorg is far away, we are somewhere between Xorg 7.0 and 7.1 plus many backports. I will try to backport XC-MISC but I am wondering if there's a mapping between the resource ids of nxagent and those of the real server. If so, we'd need an own implementation of the XC-MISC extension.

**Comment by uli42** [comment]
Now, here's a new test program which takes 40s on my rather old Core2Duo to trigger the bug:
```
#include <X11/Xlib.h>

void main(void) {
  Window w;
  Display *d = XOpenDisplay(NULL);
  if (d)
    for(;;)
      {
        w = XCreateSimpleWindow(d, RootWindow(d, 0), 10, 10, 100, 100, 1,
                                BlackPixel(d, 0), WhitePixel(d, 0));
        if (w)
          XDestroyWindow(d, w);
      }
}
```
@mbecroft: can you confirm it triggers the bug at your side, too?

**Comment by uli42** [comment]
Here's a diff of the extension in 3.5.0.32 vs. 3.5.99:
```
$ diff -y extensions_3.5.0 extensions_3.5.99 
    BIG-REQUESTS                                                    BIG-REQUESTS
    DAMAGE                                                          DAMAGE
    DEC-XTRAP                                                 <
    DOUBLE-BUFFER                                                   DOUBLE-BUFFER
    DPMS                                                            DPMS
    Extended-Visual-Information                               <
    GLX                                                             GLX
                                                              >     MIT-SCREEN-SAVER
    MIT-SHM                                                         MIT-SHM
    MIT-SUNDRY-NONSTANDARD                                    <
    RANDR                                                           RANDR
    RECORD                                                          RECORD
    RENDER                                                          RENDER
    SECURITY                                                        SECURITY
    SGI-GLX                                                         SGI-GLX
    SHAPE                                                           SHAPE
    SYNC                                                            SYNC
    TOG-CUP                                                   <
    X-Resource                                                      X-Resource
    XC-APPGROUP                                               <
    XC-MISC                                                   <
    XFIXES                                                          XFIXES
    XFree86-Bigfont                                                 XFree86-Bigfont
    XINERAMA                                                        XINERAMA
    XInputExtension                                                 XInputExtension
    XKEYBOARD                                                       XKEYBOARD
    XTEST                                                           XTEST
    XVideo                                                          XVideo
```

Dropping XC-MISC was really a bad idea... Throwing out all the others was ok, however.

**Comment by uli42** [comment]
Turns out it was not a "bad idea" but a simple flaw while dropping other, obsolete extensions. The fix is a one-liner...
